File: frontend/src/types/compliance.ts
export interface CMMCDomainStatus {
  domain: string;
  domain_name: string;
  total_controls: number;
  met_controls: number;
  percentage: number;
}

export interface CMMCStatus {
  total_controls: number;
  met_controls: number;
  score_percentage: number;
  target_level: number;
  domains: CMMCDomainStatus[];
}

export interface NISTControlFamily {
  family_id: string;
  name: string;
  total_controls: number;
  implemented: number;
  partial: number;
  not_implemented: number;
}

export interface NISTOverview {
  framework: string;
  total_families: number;
  families: NISTControlFamily[];
  overall_coverage: number;
}

export interface DataPrivacyInfo {
  data_handling: string[];
  encryption: string[];
  access_controls: string[];
  data_retention: string[];
  certifications: string[];
}

export interface TrustCenterPolicy {
  allow_ai_requirement_analysis: boolean;
  allow_ai_draft_generation: boolean;
  require_human_review_for_submission: boolean;
  share_anonymized_product_telemetry: boolean;
  retain_prompt_logs_days: number;
  retain_output_logs_days: number;
}

export interface TrustCenterPolicyUpdate {
  allow_ai_requirement_analysis?: boolean;
  allow_ai_draft_generation?: boolean;
  require_human_review_for_submission?: boolean;
  share_anonymized_product_telemetry?: boolean;
  retain_prompt_logs_days?: number;
  retain_output_logs_days?: number;
}

export interface TrustCenterRuntimeGuarantees {
  model_provider: string;
  processing_mode: string;
  provider_training_allowed: boolean;
  provider_retention_hours: number;
  no_training_enforced: boolean;
}

export interface TrustCenterEvidenceItem {
  control: string;
  status: "enforced" | "warning" | "configured";
  detail: string;
}

export interface TrustCenterProfile {
  organization_id: number | null;
  organization_name: string | null;
  can_manage_policy: boolean;
  policy: TrustCenterPolicy;
  runtime_guarantees: TrustCenterRuntimeGuarantees;
  evidence: TrustCenterEvidenceItem[];
  updated_at: string;
}

export interface ComplianceAuditSummary {
  total_events: number;
  events_last_30_days: number;
  by_type: Record<string, number>;
  compliance_score: number;
}

export interface ComplianceReadinessProgram {
  id: string;
  name: string;
  status: string;
  percent_complete: number;
  next_milestone: string;
}

export interface ComplianceReadiness {
  programs: ComplianceReadinessProgram[];
  last_updated: string;
}

export interface ComplianceReadinessCheckpoint {
  checkpoint_id: string;
  program_id: string;
  title: string;
  status: string;
  target_date: string;
  owner: string;
  third_party_required: boolean;
  evidence_items_ready: number;
  evidence_items_total: number;
  evidence_source?: "static" | "registry" | null;
  evidence_last_updated_at?: string | null;
  assessor_signoff_status?: "pending" | "approved" | "rejected" | null;
  assessor_signoff_by?: string | null;
  assessor_signed_at?: string | null;
}

export interface ComplianceReadinessCheckpointSnapshot {
  checkpoints: ComplianceReadinessCheckpoint[];
  generated_at: string;
}

export interface ComplianceTrustMetrics {
  generated_at: string;
  window_days: number;
  checkpoint_evidence_completeness_rate: number | null;
  checkpoint_signoff_completion_rate: number | null;
  trust_export_success_rate_30d: number | null;
  trust_export_successes_30d: number;
  trust_export_failures_30d: number;
  step_up_challenge_success_rate_30d: number | null;
  step_up_challenge_successes_30d: number;
  step_up_challenge_failures_30d: number;
  trust_ci_pass_rate_30d: number | null;
}

export interface ComplianceRegistryEvidenceItem {
  id: number;
  user_id: number;
  title: string;
  evidence_type: string;
  description: string | null;
  file_path: string | null;
  url: string | null;
  collected_at: string;
  expires_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface ComplianceCheckpointEvidenceItem {
  link_id: number;
  checkpoint_id: string;
  evidence_id: number;
  title: string;
  evidence_type: string;
  description: string | null;
  file_path: string | null;
  url: string | null;
  collected_at: string;
  expires_at: string | null;
  status: "submitted" | "accepted" | "rejected";
  notes: string | null;
  reviewer_user_id: number | null;
  reviewer_notes: string | null;
  reviewed_at: string | null;
  linked_at: string;
}

export interface ComplianceCheckpointEvidenceCreate {
  evidence_id: number;
  status?: "submitted" | "accepted" | "rejected";
  notes?: string;
}

export interface ComplianceCheckpointEvidenceUpdate {
  status?: "submitted" | "accepted" | "rejected";
  reviewer_notes?: string;
}

export interface ComplianceCheckpointSignoff {
  checkpoint_id: string;
  status: "pending" | "approved" | "rejected";
  assessor_name: string;
  assessor_org: string | null;
  notes: string | null;
  signed_by_user_id: number | null;
  signed_at: string | null;
  updated_at: string;
}

export interface ComplianceCheckpointSignoffWrite {
  status: "pending" | "approved" | "rejected";
  assessor_name: string;
  assessor_org?: string;
  notes?: string;
}

export interface GovCloudMigrationPhase {
  phase_id: string;
  title: string;
  status: string;
  target_date: string;
  owner: string;
  exit_criteria: string[];
}

export interface GovCloudDeploymentProfile {
  program_id: string;
  provider: string;
  status: string;
  target_regions: string[];
  boundary_services: string[];
  identity_federation_status: string;
  network_isolation_status: string;
  data_residency_status: string;
  migration_phases: GovCloudMigrationPhase[];
  updated_at: string;
}

export interface SOC2ControlDomainStatus {
  domain_id: string;
  domain_name: string;
  controls_total: number;
  controls_ready: number;
  percent_complete: number;
  owner: string;
}

export interface SOC2Milestone {
  milestone_id: string;
  title: string;
  status: string;
  due_date: string;
  owner: string;
  evidence_ready: boolean;
  notes: string;
}

export interface SOC2Readiness {
  program_id: string;
  name: string;
  status: string;
  audit_firm_status: string;
  observation_window_start: string;
  observation_window_end: string;
  overall_percent_complete: number;
  domains: SOC2ControlDomainStatus[];
  milestones: SOC2Milestone[];
  updated_at: string;
}

File: backend/app/schemas/compliance.py
"""
Compliance dashboard and trust-center schemas.
"""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator


class DataPrivacyInfo(BaseModel):
    data_handling: list[str]
    encryption: list[str]
    access_controls: list[str]
    data_retention: list[str]
    certifications: list[str]


class ComplianceReadinessProgram(BaseModel):
    id: str
    name: str
    status: str
    percent_complete: int = Field(ge=0, le=100)
    next_milestone: str


class ComplianceReadinessResponse(BaseModel):
    programs: list[ComplianceReadinessProgram]
    last_updated: str


class ComplianceReadinessCheckpoint(BaseModel):
    checkpoint_id: str
    program_id: str
    title: str
    status: str
    target_date: datetime
    owner: str
    third_party_required: bool
    evidence_items_ready: int = Field(ge=0)
    evidence_items_total: int = Field(ge=0)
    evidence_source: Literal["static", "registry"] | None = None
    evidence_last_updated_at: datetime | None = None
    assessor_signoff_status: Literal["pending", "approved", "rejected"] | None = None
    assessor_signoff_by: str | None = None
    assessor_signed_at: datetime | None = None


class ComplianceReadinessCheckpointResponse(BaseModel):
    checkpoints: list[ComplianceReadinessCheckpoint]
    generated_at: datetime


class ComplianceCheckpointEvidenceCreate(BaseModel):
    evidence_id: int
    status: Literal["submitted", "accepted", "rejected"] | None = None
    notes: str | None = None


class ComplianceCheckpointEvidenceUpdate(BaseModel):
    status: Literal["submitted", "accepted", "rejected"] | None = None
    reviewer_notes: str | None = None


class ComplianceCheckpointEvidenceRead(BaseModel):
    link_id: int
    checkpoint_id: str
    evidence_id: int
    title: str
    evidence_type: str
    description: str | None = None
    file_path: str | None = None
    url: str | None = None
    collected_at: datetime
    expires_at: datetime | None = None
    status: Literal["submitted", "accepted", "rejected"]
    notes: str | None = None
    reviewer_user_id: int | None = None
    reviewer_notes: str | None = None
    reviewed_at: datetime | None = None
    linked_at: datetime


class ComplianceCheckpointSignoffWrite(BaseModel):
    status: Literal["pending", "approved", "rejected"]
    assessor_name: str
    assessor_org: str | None = None
    notes: str | None = None


class ComplianceCheckpointSignoffRead(BaseModel):
    checkpoint_id: str
    status: Literal["pending", "approved", "rejected"]
    assessor_name: str
    assessor_org: str | None = None
    notes: str | None = None
    signed_by_user_id: int | None = None
    signed_at: datetime | None = None
    updated_at: datetime


class GovCloudMigrationPhase(BaseModel):
    phase_id: str
    title: str
    status: str
    target_date: datetime
    owner: str
    exit_criteria: list[str]


class GovCloudDeploymentProfile(BaseModel):
    program_id: str
    provider: str
    status: str
    target_regions: list[str]
    boundary_services: list[str]
    identity_federation_status: str
    network_isolation_status: str
    data_residency_status: str
    migration_phases: list[GovCloudMigrationPhase]
    updated_at: datetime


class SOC2ControlDomainStatus(BaseModel):
    domain_id: str
    domain_name: str
    controls_total: int = Field(ge=0)
    controls_ready: int = Field(ge=0)
    percent_complete: int = Field(ge=0, le=100)
    owner: str


class SOC2Milestone(BaseModel):
    milestone_id: str
    title: str
    status: str
    due_date: datetime
    owner: str
    evidence_ready: bool
    notes: str


class SOC2ReadinessResponse(BaseModel):
    program_id: str
    name: str
    status: str
    audit_firm_status: str
    observation_window_start: datetime
    observation_window_end: datetime
    overall_percent_complete: int = Field(ge=0, le=100)
    domains: list[SOC2ControlDomainStatus]
    milestones: list[SOC2Milestone]
    updated_at: datetime


class TrustCenterPolicy(BaseModel):
    allow_ai_requirement_analysis: bool
    allow_ai_draft_generation: bool
    require_human_review_for_submission: bool
    share_anonymized_product_telemetry: bool
    retain_prompt_logs_days: int = Field(ge=0, le=30)
    retain_output_logs_days: int = Field(ge=0, le=365)


class TrustCenterPolicyUpdate(BaseModel):
    allow_ai_requirement_analysis: bool | None = None
    allow_ai_draft_generation: bool | None = None
    require_human_review_for_submission: bool | None = None
    share_anonymized_product_telemetry: bool | None = None
    retain_prompt_logs_days: int | None = Field(default=None, ge=0, le=30)
    retain_output_logs_days: int | None = Field(default=None, ge=0, le=365)

    @field_validator("retain_prompt_logs_days", "retain_output_logs_days")
    @classmethod
    def _ensure_non_negative(cls, value: int | None) -> int | None:
        if value is not None and value < 0:
            raise ValueError("Retention days must be >= 0")
        return value


class TrustCenterRuntimeGuarantees(BaseModel):
    model_provider: str
    processing_mode: str
    provider_training_allowed: bool
    provider_retention_hours: int
    no_training_enforced: bool


class TrustCenterEvidenceItem(BaseModel):
    control: str
    status: Literal["enforced", "warning", "configured"]
    detail: str


class TrustCenterProfile(BaseModel):
    organization_id: int | None = None
    organization_name: str | None = None
    can_manage_policy: bool
    policy: TrustCenterPolicy
    runtime_guarantees: TrustCenterRuntimeGuarantees
    evidence: list[TrustCenterEvidenceItem]
    updated_at: datetime
